README.md
# ffmpeg-android

* FFmpeg for Android with libvpx, x264 and lame options. The scripts in this repository are configured to build:
<table>
<thead>
<tr><td>library</td><td>version</td><td>platform support</td><td>arch support</td></tr>
</thead>
<tr><td>ffmpeg</td><td>4.4</td><td>android</td><td>armeabi-v7a arm64-v8a x86 x86_64</td></tr>
<tr><td>x264</td><td>163</td><td>android</td><td>armeabi-v7a arm64-v8a x86 x86_64</td></tr>
<tr><td>libvpx</td><td>1.10.0</td><td>android</td><td>armeabi-v7a arm64-v8a x86 x86_64</td></tr>
<tr><td>lame</td><td>3.100</td><td>android</td><td>armeabi-v7a arm64-v8a x86 x86_64</td></tr>
</table>

## Supported android ABI's
* armeabi*
* armeabi-v7a
* arm64-v8a
* x86
* x86_64
* mips*
* mips64*

Note: *-Deprecated in android ndk-r16. Will be removed in ndk-r17.<br/>
see https://developer.android.com/ndk/guides/abis.html#Supported ABIs <br/>
ffmpeg-andorid releases > v1.5.0 will only verify ABIS=("armeabi-v7a" "arm64-v8a" "x86" "x86_64")

## Instructions
* Set environment ANDROID_NDK variable; not all ndk releases work with all modules/ABIS combinations
  - export ANDROID_NDK={Android NDK Base Path}<br/>
     e.g. export ANDROID_NDK=/opt/android/android-ndk-r17c (recommended);
  - Note: android-ndk-r18b will have problem in some module builds e.g. lame
* If necessary, fetch and update all libraries source (before build);
  - review and edit ./init_update_libs.sh file for your desired modules' versions, then run
  - ./init_update_libs.sh
  - The default codec module extracted sub-directories are ffmpeg, libvpx, x264 and lame
  - Optionally, go to the respective directory for each sub-module, and execute ./configure without option;<br/>
     The configure process may list any missing sdk build tools, please install before continue
  - The actual configure with options for each submodule build is done in each submodule _\<module>_build.sh build script
* Edit ./ffmpeg-android_build.sh \[#1] and ./_settings.sh \[#2];<br/>
  a. remove any of the codec sub-modules or architectures you wish to be excluded from the build<br/>
  b. add "ffmpeg" only to \[#1] if you want to build standalone ffmpeg without inclusion of the sub-modules integration<br/>
  c. the default values are defined as:
  - MODULES=("vpx" "x264" "lame") \[#1]
  - ABIS=("armeabi-v7a" "arm64-v8a" "x86" "x86_64") \[#2]
* To support 64-bit libraries built, ensure _settings.sh#ANDROID_API=21 (min API for 64-bit library build).<br/>
  Note: application.mk in Android Studio i.e. APP_PLATFORM := android-21 i.e. both must use the same API
* Use the following commands to compile ffmpeg for all or custom defined modules/ABI's combinations
  - ./ffmpeg-android_build.sh<br/>: (combinaiton as specified in the \[#1] and \[#2])
  - ./ffmpeg-android_build.sh armeabi-v7a<br/>: (i.e. selected cpu with all predefined codec modules)
  - ./ffmpeg-android_build.sh armeabi-v7a x264<br/>: (i.e. selected cpu with only x264 codec module)
* All the generated static/shared libraries and includes files are installed in ./jni/\<MODULE>/android/\<ABI>.

## Linking with versioned shared library in Android NDK
* Note: Manual changes as outlined below are not further required.<br/>
  the latest _vpx_build.sh has included the scripts to perform the following steps automatically.
* Android has an issue with loading versioned .so shared libraries e.g. x264:
* Causing error during run: java.lang.UnsatisfiedLinkError: dlopen failed: library "libx264.so.147" not found
* Perform the following patches, if you want to link with shared .so libraries for x264.
  1. use GHex to change file content "libx264.so.147" to "libx264_147.so"
  2. change filename from libx264.so.147 to libx264_147.so
  3. Note: the .so filename must match with the file changed content


## Verification Status
* The scripts has been verified working with the following configurations:
  - NDK version: ndk-r17c (build may failed with lower or higher versions - see below)
  - ABIS: armeabi-v7a, arm64-v8a, x86, x86_64
  - MODULES (with applied patches): ffmpeg-v4.4, libvpx-v1.10.0, x264-v163, lame-v3.100
  - ANDROID_API: 21

* x264 (v163, v161, v160, v157 and v152):
  - x264 option: --disable-asm<br/>
    Must exclude the option for arm64. Used by configure, config.mak and Makefile to define AS and to compile required *.S assembly files.<br/>
    Otherwise will have undefined references e.g. x264_8_... x264_10...<br/>
    However must include the option for x86 and x86_64; otherwise have relocate text, requires dynamic R_X86_64_PC32 etc when use in aTalk
  - ndk-r18b, ndk-r17c, ndk-16b, ndk-r15c (build all)

* libvpx (v1.10.0, v.1.8.2, v1.8.0):
  - libvpx configure.sh needs patches to correctly build the arm64 with NDK standalone toolchains.
  - Valid for libvpx v1.8.0 only, option removed for libvpx v1.8.2+:<br/>
    When --sdk-path is specified, libvpx configure uses SDK toolchains compiler (gcc/g++);
    * ndk-r18b, ndk-r17c, ndk-r16b:<br/>
      To avoid missing stdlib.h and other errors, need to include the following two options for SDK toolchains:<br/>
      --extra-cflags="-isystem ${NDK}/sysroot/usr/include/${NDK_ABIARCH} -isystem ${NDK}/sysroot/usr/include"<br/>
      --libc=${NDK_SYSROOT} => use standalone toolchains directory.<br/>
      libvpx v1.8.0 configure.sh has a problem configure this with SDK properly
    * ndk-r18b: gcc option has been removed<br/>
      build failed with: /home/cmeng/workspace/ndk/ffmpeg-android/toolchain-android/bin/aarch64-linux-android-ld: cannot find -lgcc
  - libvpx v1.8.0: When using standalone toolchains, i.e. omit --sdk-path;
    * ndk-r18b, ndk-r17c, ndk-r16b: <br/>
      Build ok with ABIS=("arm64-v8a" "x86" "x86_64") but not "armeabi-v7a" and failed with:<br/>
    /tmp/vpx-conf-4350-25363.o(.ARM.exidx.text.main+0x0): error: undefined reference to '__aeabi_unwind_cpp_pr0'
  - For libvpx v1.8.2+
    * ndk-r18b, ndk-r17c, ndk-r16b: <br/>
      All built OK.

* lame (v3.1000):
  - ndk-r18b - failed as lame need C++ instead of clang compiler
  - ndk-r17b - build all and can integrate with ffmpeg
  - ndk-r16b, ndk-r15c (build all)
  - PREFIX must use absolution path

* ffmpeg (v4.4):
  - Must include option --disable-asm for x86 and x86_64, otherwise <br/>
    libavcodec/x86/cabac.h:193:9: error: inline assembly requires more registers than available<br/>
    ffmpeg (v1.0.10) => must also include this option for arm/arm64 build, otherwise errors during compilation 
  - ndk-r18b, ndk-r17c => give error on:<br/>
    libavdevice/v4l2.c:135:9: error: assigning to 'int (*)(int, unsigned long, ...)' from incompatible type '<overloaded function type>'
        SET_WRAPPERS();<br/>
    add to CFLAG: -DBIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD; see https://github.com/tanersener/mobile-ffmpeg/issues/48
  - ndk-r16b or lower => gives error on:<br/>
    libavformat/udp.c:290:28: error: member reference base type '__be32' (aka 'unsigned int') is not a structure or union<br/>
        mreqs.imr_multiaddr.s_addr = ((struct sockaddr_in *)addr)->sin_addr.s_addr;
  
* Note:
  - NDK verification status (build with clang/clang++ and API-21 unless otherwise specified):
  - Recommendation: use ndk-r17c (work for all) and API-21 unless you have other considerations.
  - clang support needs min ndk-r12c, but may not necessary works for all ABIS (64-bit built has problem)
  - Change _settings.sh with clang=>gcc and clang++=>g++ if you need lower ndk version
  - NDK >r18b has obsoleted the support for the gcc/g++

## Help:
* Set up Linux/Ubuntu development environment with the below build tools
  - sudo apt-get --quiet --yes install build-essential git autoconf libtool pkg-config gperf gettext yasm python-lxml
* Patches for Sub-module
  - ./ffmpeg-android_build.sh includes the patches for the sub-modules 
  - ffmpeg-android_patch.sh applies patches to the relevant sub-module with patch files from ./pathes directory
  - edit these files to include additional patches if required.
* Configuration failed
  - You may encounter this problem during the codec sub-module build, navigate to the respective sub-module directory:
    - Issue the command line i.e. configure --help to check for the available options 
    - Refer to config.log of the sub-module for more info.
    - Refer to the configure file for more information on CPU types supported
    - Edit the script file to make the necessary modifications based on help and errors found in config.log
* During sub-module built, you may encountered compilation or linker errors, changing ndk version <br/>may help to resolve the issues.
  However it may create new problems in another areas.
  
* Utilize Modern Compiler Flags to Address Potential Security Issues
  - Stack execution protection:                    LDFLAGS="-z noexecstack" 
  - Data relocation and protection (RELRO):        LDLFAGS="-z relro -z now" 
  - Stack-based Buffer Overrun Detection:          CFLAGS=”-fstack-protector-strong”
                                                   if using GCC 4.9 or newer, otherwise CFLAGS="-fstack-protector"
  - Position Independent Execution (PIE)           CFLAGS="-fPIE -fPIC" LDFLAGS="-pie" (PIE for executables only)
  - Fortify source:                                CFLAGS="-O2 -D_FORTIFY_SOURCE=2"
  - Format string vulnerabilities:                 CFLAGS="-Wformat -Wformat-security"

* All the information provided here are by trial and error; compiling and linkage are very sensitive to the<br/>
  modules version, configure options, and the NDK version used. You are advice to build without any changes<br/>
  to the scripts provided here to confirm working, before attempt to make any modifications.
   
## Note:
* The scripts in this folder are not compatible with Unified Headers:<br/>
See https://android.googlesource.com/platform/ndk/+/master/docs/UnifiedHeaders.md#supporting-unified-headers-in-your-build-system

* Both the android NDK, ffmpeg and its sub-modules are in continous update, it is likely that <br/>
some of the defined ABI's settings may need to be tweaked as submodule configure have new changes.

Please refer to the following sites which may offer solution for problems you may experience.
* https://ffmpeg.org/pipermail/ffmpeg-user/2016-January/030202.html
* https://www.mail-archive.com/ffmpeg-devel@ffmpeg.org/msg62644.html
* http://alientechlab.com/how-to-build-ffmpeg-for-android/
* https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures
* https://github.com/google/ExoPlayer/issues/3520 (VP9 builds failure with android-ndk-r16 #3520)
* https://github.com/android-ndk/ndk/issues/190#issuecomment-375164450 (unknown type name __uint128_t on ndk-build #190)
* https://android.googlesource.com/platform/bionic/+/master/docs/32-bit-abi.md
* https://github.com/android-ndk/ndk/issues/477 (mmap causes compile errors on r15c (unknown identifier) #477)
* https://github.com/android-ndk/ndk/issues/503 (implicit declaration of function 'mmap' #503)
* https://github.com/nodejs/node/issues/18671 (Utilize Modern Compiler Flags to Address Potential Security Issues #18671)
* https://android.googlesource.com/platform/external/libvpx/+/ca30a60d2d6fbab4ac07c63bfbf7bbbd1fe6a583 (Add visibility="protected" attribute for global variables referenced in asm files.)


## License

    ffmpeg, android static library for aTalk VoIP and Instant Messaging client
    
    Copyright 2014 Eng Chong Meng
        
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
       http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.




Changelog
# Entries are sorted chronologically from oldest to youngest within each release,
# releases are sorted from youngest to oldest.
#=============================================================
version 1.6.8 (2021/07/19)
- Clean up scripts for all submodules build
- Change method to determine ffmpeg version
- include version check in init_update_libs.sh before taking action
- include patch for libvpx-1.10.0
- Scripts are verified for all latest version libraries only; see init_update_libs.sh

#=============================================================
version 1.6.7 (2021/05/17)
- Clean up scripts for ffmpeg v4.4 build
- Clean up scripts for x264 v161.3049 build

#=============================================================
version 1.6.6 (2021/04/28)
- Clean up scripts for libvpx v1.10.0 build

#=============================================================
version 1.6.5 (2020/07/25)
- add amr codec
- use option --enable-libopencore-amrwb --enable-libvo-amrwbenc --enable-version3

#=============================================================
version 1.6.4 (2020/07/17)
- Clean up scripts for libvpx v1.8.2 build
- Update lame patch file to use configure file for version extraction
- Include solution for ffmpeg 4.1.1 when compiled with NDK r18b

#=============================================================
version 1.6.3 (2019/05/24)
- use option --enable-runtime-cpu-detect for x86/x86_64 platforms in vpx library build to fix runtime problem

#=============================================================
version 1.6.2 (2019/03/29)
- Add support for ffmpeg-1.0.10 build with clang (generated *.a libraries seems large)
- Include patches for ffmpeg-1.0.10
- Include --disable-asm for arm/arm64 for ffmpeg-1.0.10 when using clang due to errors

#=============================================================
vesrion 1.6.1 (2019/03/28)
- Add new patches for libvpx fixing: relocation R_386_GOTOFF against preemptible symbol... (valid for v1.8.0, v1.7.0 & v1.6.1+)
- Add new patches for libvpx fixing: sysroot for arm64-v8a
- Add option --enable-pic (required when using SDK toolchains)
- Add option to build standalone ffmpeg (aTalk) without the inclusion of the codec sub-modules
- Add option --disable-postproc to ffmpeg build
- Add dependency for options that are ABIS depends in _<MODULE>_build.sh scripts
- Clean up README.md information
- Display module version in applying patches and building process

#=============================================================
version 1.6.0 (2019/03/18)
- Use ffmpeg-4.1.1, libvpx-1.8.0, x264-157 and lame-3.100  for verification
- Upate patches to include libvpx and x264 etc
- Update _vpx_build.sh to use SDK or Standalone toolchains (see _vpx_build.sh for details)
- Change prefix to each sub-modules sub-directory for compatibility with aTalk
- prefix all sub-modules script file with '_' for better clarifications
- Cleanup all build scripts and readme file

#=============================================================
version 1.5.0 (2018/07/24):
- clean up build scripts
- Library ffmpeg use version 4.0.2
- Library libvpx use version 1.7.x (master-20180727)

#=============================================================
version 1.4.0 (2018/03/24):
- move LDFLAGS to ffmpeg_build.sh - only requires for final ffmpeg linkage
- libvpx still has configure error for Target=arm64-android-gcc

#=============================================================
version 1.3.0 (2018/03/21):
- cleanup and add libvpx options to _settings.sh
- include libvpx build (ndk-r15c only and r16b has errors - see vpx_build.sh)
- libvpx for arm64-v8a needs neon options, however libvpx.a not compatible with ffmpeg
- remove -nostadlib option in LDFLAGS, lame check stdlib for linkage

#=============================================================
version 1.2.0 (2018/03/18):
- update init_update_libs.sh to use lame-3.100 for aarch64 support; arm64-v8a libx.a has own undefined references
- change mips/mips64 cpu to use p5600/i6400 - ones that have been verified working
- ABI mips64 : -march=mips64r6 accepts by clang50 but complain by ffmpeg
  -march=i6400 works for ffmpeg but complain by clang50; omit it in CFLAG - works for both

#=============================================================
version 1.1.0 (2018/03/14):
- clean up _settings.sh export parameters
- default to use API-21 for 64bit build
- use same PREFIX directory for all compiled sub-modules static libraries
- change FFMPEG_PKG_CONFIG to point to PREFIX directory
- apply patches for ffmpeg files (#undefine B0)

#=============================================================
version 1.0.0 (2018/03/11):
- initial release


_settings.sh
#!/bin/bash
#
# Copyright 2016 Eng Chong Meng
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Uncomment the line below to see all script echo to terminal
# set -x

# When use --sdk-path option for libvpx v1.8.0; must use android-ndk-r17c or lower
# May use android-ndk-r18b" - libvpx v1.10.0 and libvpx v1.8.2 are working with r18b without error
# lame needs android-ndk-r17c
export ANDROID_NDK=/opt/android/android-ndk-r17c
if [[ -z $ANDROID_NDK ]] || [[ ! -d $ANDROID_NDK ]] ; then
	echo "You need to set ANDROID_NDK environment variable, exiting"
	echo "Use: export ANDROID_NDK=/your/path/to/android-ndk-rxx"
	echo "e.g.: export ANDROID_NDK=/opt/android/android-ndk-r17c"
	exit 1
fi

set -u

# Never mix two api level to build static library for use on the same apk.
# Set to API:21 for aTalk 64-bit architecture support
# Does not build 64-bit arch if ANDROID_API is less than 21 i.e. the minimum supported API level for 64-bit.
ANDROID_API=21

# set STANDALONE_TOOLCHAINS to 0: SDK toolchains OR 1: standalone toolchains
STANDALONE_TOOLCHAINS=1;

# Built with command i.e. ./build-ffmpeg4android.sh or following with parameter [ABIS(x)]
# Create custom ABIS or uncomment to build all supported abi for ffmpeg.
# Do not change naming convention of the ABIS; see:
# https://developer.android.com/ndk/guides/abis.html#Native code in app packages
# ABIS=("armeabi" "armeabi-v7a" "arm64-v8a" "x86" "x86_64" "mips" "mips64")

# Android recommended architecture support; others are deprecated
ABIS=("armeabi-v7a" "arm64-v8a" "x86" "x86_64")

BASEDIR=`pwd`
TOOLCHAIN_PREFIX=${BASEDIR}/toolchain-android

#===========================================
# Do not proceed further on first call without the required 2 parameters
[[ $# -lt 2 ]] && return

NDK=${ANDROID_NDK}
HOST_NUM_CORES=$(nproc)

# https://gcc.gnu.org/onlinedocs/gcc-4.9.1/gcc/Optimize-Options.html
# Note: vpx with ABIs x86 and x86_64 build has error with option -fstack-protector-all

# Note: final libraries built is 20~33% bigger in size when below additional options are specified
# CFLAGS_="-DANDROID -fpic -fpie -ffunction-sections -funwind-tables -fstack-protector -fno-strict-aliasing -fno-strict-overflow -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2"
# -DBIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD: see https://github.com/tanersener/mobile-ffmpeg/issues/48
CFLAGS_="-DANDROID -fpic -fpie -DBIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD"

# Enable report-all for earlier detection of errors instead at later stage
# /home/cmeng/workspace/ndk/vpx-android/armeabi-v7a-android-toolchain/bin/arm-linux-androideabi-ld: -Wl,-z,defs -Wl,--unresolved-symbols=report-all: unknown option
# Not compatible with libvpx v.1.8.2
# LDFLAGS_="-Wl,-z,defs -Wl,--unresolved-symbols=report-all"
LDFLAGS_=""

# Do not modify any of the NDK_ARCH, CPU and -march unless you are very sure.
# The settings are used by <ARCH>-linux-android-gcc and submodule configure
# https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures
# ${NDK}/toolchains/llvm/prebuilt/...../include llvm/ARMTargetParser.def etc
# NDK-ARCH - should be one from $ANDROID_NDK/platforms/android-$API/arch-* [arm / arm64 / mips / mips64 / x86 / x86_64]"
# https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html

case $1 in
  # Deprecated in r16. Will be removed in r17
  armeabi)
    CPU='armv5'
    HOST='arm-linux'
    NDK_ARCH="arm"
    NDK_ABIARCH='arm-linux-androideabi'
    CFLAGS="${CFLAGS_} -march=${CPU} -mthumb -finline-limit=64"
    ASFLAGS=""
  ;;
  # https://gcc.gnu.org/onlinedocs/gcc-4.9.4/gcc/ARM-Options.html#ARM-Options
  armeabi-v7a)
    CPU='armv7-a'
    HOST='arm-linux'
    NDK_ARCH='arm'
    NDK_ABIARCH='arm-linux-androideabi'
    # clang70: warning: -Wl,--fix-cortex-a8: 'linker' input unused [-Wunused-command-line-argument]
    # CFLAGS="${CFLAGS_} -Wl,--fix-cortex-a8 -march=${CPU} -mfloat-abi=softfp -mfpu=neon -mtune=cortex-a8 -mthumb -D__thumb__"
    CFLAGS="${CFLAGS_} -Os -march=${CPU} -mfloat-abi=softfp -mfpu=neon -mtune=cortex-a8 -mthumb -D__thumb__"
    LDFLAGS="${LDFLAGS_} -march=${CPU}"  # -Wl,--fix-cortex-a8" not valid option
    ASFLAGS=""

    # 1. -march=${CPU} flag targets the armv7 architecture.
    # 2. -mfloat-abi=softfp enables hardware-FPU instructions while ensuring that the system passes
    #     floating-point parameters in core registers, which is critical for ABI compatibility
    # 3. -mfpu=neon setting forces the use of VFPv3-D32, per the ARM specifications
    # 4. -mthumb forces the generation of 16-bit Thumb-2 instructions (Thumb-1 for armeabi).
    #     If omitted, the toolchain will emit 32-bit ARM instructions.
    # 5. -Wl,--fix-cortex-a8 is required as a workaround for a CPU bug in some Cortex-A8 implementation
    #     (x264 flags as warning) Standalone toolchains does not accept this option. SDK toolchains (gcc/cg++) is ok.
    #    /home/cmeng/workspace/ndk/ffmpeg-android/toolchain-android/bin/arm-linux-androideabi-ld: -Wl,--fix-cortex-a8: unknown option
    # LDFLAGS="-Wl,--fix-cortex-a8"

    # arm v7vfpv3
    # CFLAGS="${CFLAGS_} -march=${CPU} -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb"

    # arm v7 + neon (neon also include vfpv3-32)
    # CFLAGS="${CFLAGS_} -march=${CPU} -mfloat-abi=softfp -mfpu=neon -mtune=cortex-a8 -mthumb -D__thumb__"
  ;;
  arm64-v8a)
    # Valid cpu = armv8-a cortex-a35, cortex-a53, cortec-a57 etc. but -march=armv8-a is required
    # -march valid only for ‘armv8-a’, ‘armv8.1-a’, ‘armv8.2-a’, ‘armv8.3-a’ or ‘armv8.4-a’
    #  or native (only armv8-a is valid for lame build).
    CPU='cortex-a57'
    HOST='aarch64-linux'
    NDK_ARCH='arm64'
    NDK_ABIARCH='aarch64-linux-android'
    CFLAGS="${CFLAGS_} -O3 -march=armv8-a"
      # Supported emulations: aarch64linux aarch64elf aarch64elf32 aarch64elf32b aarch64elfb armelf armelfb
      # aarch64linuxb aarch64linux32 aarch64linux32b armelfb_linux_eabi armelf_linux_eabi
	  #-march=armv8-a or arch64linux: all are not valid for libvpx v1.8.2/v1.10.0 build with standalone toolchains
    LDFLAGS="${LDFLAGS_}"
    ASFLAGS=""
  ;;
  x86)
    CPU='i686'
    HOST='i686-linux'
    NDK_ARCH='x86'
    NDK_ABIARCH='i686-linux-android'
    CFLAGS="${CFLAGS_} -O3 -march=${CPU} -mtune=intel -msse3 -mfpmath=sse -m32 -fPIC"
    LDFLAGS="-m32"
    ASFLAGS="-D__ANDROID__"
  ;;
  x86_64)
    CPU='x86-64' 
    HOST='x86_64-linux'
    NDK_ARCH='x86_64'
    NDK_ABIARCH='x86_64-linux-android'
    CFLAGS="${CFLAGS_} -O3 -march=${CPU} -mtune=intel -msse4.2 -mpopcnt -m64 -fPIC"
    LDFLAGS=""
    ASFLAGS="-D__ANDROID__"
  ;;

  # MIPS is deprecated in NDK r16 and will be removed in r17.
  # https://gcc.gnu.org/onlinedocs/gcc/MIPS-Options.html#MIPS-Options
  # https://en.wikipedia.org/wiki/List_of_MIPS_architecture_processors
  mips)
    # unknown cpu - optimization disable
    CPU='p5600'
    HOST='mips-linux'
    NDK_ARCH='mips'
    NDK_ABIARCH="mipsel-linux-android"
    CFLAGS="${CFLAGS_} -EL -march=p5600 -mhard-float"
    ASFLAGS=""
  ;;
  mips64)
    # -march=mips64r6 works for clangs but complain by ffmpeg (use -march=${CPU}), reverse effect when -march=i6400 - so omit it in CFLAG works for both
    CPU='i6400'
    HOST='mips64-linux'
    NDK_ARCH='mips64'
    NDK_ABIARCH='mips64el-linux-android'
    CFLAGS="${CFLAGS_} -EL -mfp64 -mhard-float"
    ASFLAGS=""
  ;;
esac

# Create standalone toolchains for the specified architecture - use .py instead of the old .sh
# However for ndk--r19b => Instead use:
#    $ ${NDK}/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android21-clang++ src.cpp
if [[ ${STANDALONE_TOOLCHAINS} == 1 ]]; then
  TOOLCHAIN_PREFIX=${BASEDIR}/toolchain-android
  NDK_SYSROOT=${TOOLCHAIN_PREFIX}/sysroot
  CC_=clang
  CXX_=clang++

  if [[ ! -e ${TOOLCHAIN_PREFIX}/${NDK_ABIARCH} ]]; then
    rm -rf ${TOOLCHAIN_PREFIX}

  # Create standalone toolchains for the specified architecture - use .py instead of the old .sh
  # However for ndk--r19b => Instead use:
  #    $ ${NDK}/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android21-clang++ src.cpp
  # cmeng: must ensure AS JNI uses the same STL library or "system" if specified
    [[ -d ${TOOLCHAIN_PREFIX} ]] || python ${NDK}/build/tools/make_standalone_toolchain.py \
      --arch ${NDK_ARCH} \
      --api ${ANDROID_API} \
      --stl libc++ \
      --install-dir=${TOOLCHAIN_PREFIX}
  fi
else
  TOOLCHAIN_PREFIX=${ANDROID_NDK}/toolchains/${NDK_ABIARCH}-49/prebuilt/linux-x86_64
  NDK_SYSROOT=${ANDROID_NDK}/platforms/android-${ANDROID_API}/arch-${NDK_ARCH}
  CC_=gcc
  CXX_=g++
fi

# Define the install directory of the libs and include files etc
# lame needs absolute path
# PREFIX=../jni/$2/android/$1
PREFIX=${BASEDIR}/jni/$2/android/$1

# Add the standalone toolchain to the search path.
export PATH=${TOOLCHAIN_PREFIX}/bin:$PATH
export CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/${NDK_ABIARCH}-
export CFLAGS="${CFLAGS}"
export CPPFLAGS="${CFLAGS}"
export CXXFLAGS="${CFLAGS} -std=c++11"
export ASFLAGS="${ASFLAGS}"
export LDFLAGS="${LDFLAGS} -L${NDK_SYSROOT}/usr/lib"

export AR="${CROSS_PREFIX}ar"
export AS="${CROSS_PREFIX}${CC_}"
export CC="${CROSS_PREFIX}${CC_}"
export CXX="${CROSS_PREFIX}${CXX_}"
export LD="${CROSS_PREFIX}ld"
export STRIP="${CROSS_PREFIX}strip"
export RANLIB="${CROSS_PREFIX}ranlib"
export OBJDUMP="${CROSS_PREFIX}objdump"
export CPP="${CROSS_PREFIX}cpp"
export GCONV="${CROSS_PREFIX}gconv"
export NM="${CROSS_PREFIX}nm"
export SIZE="${CROSS_PREFIX}size"
export PKG_CONFIG="${CROSS_PREFIX}pkg-config"
export PKG_CONFIG_LIBDIR=${PREFIX}/lib/pkgconfig
export PKG_CONFIG_PATH=${PREFIX}/lib/pkgconfig

echo "**********************************************"
echo "### Use NDK=${NDK}"
echo "### Use ANDROID_API=${ANDROID_API}"
echo "### Install directory: PREFIX=${PREFIX}"
echo "**********************************************"

# Undefined reference in x264 when linked with ffmpeg for arm64-v8a
# https://stackoverflow.com/questions/5555632/can-gcc-not-complain-about-undefined-references
#
#It is possible to avoid reporting undefined references - using --unresolved-symbols linker option.
#
#g++ mm.cpp -Wl,--unresolved-symbols=ignore-in-object-files
#From man ld
#
#--unresolved-symbols=method
#Determine how to handle unresolved symbols. There are four possible values for method:
#
#       ignore-all
#           Do not report any unresolved symbols.
#
#       report-all
#           Report all unresolved symbols.  This is the default.
#
#       ignore-in-object-files
#           Report unresolved symbols that are contained in shared
#           libraries, but ignore them if they come from regular object
#           files.
#
#       ignore-in-shared-libs
#           Report unresolved symbols that come from regular object
#           files, but ignore them if they come from shared libraries.  This
#           can be useful when creating a dynamic binary and it is known
#           that all the shared libraries that it should be referencing
#           are included on the linker's command line.
#The behaviour for shared libraries on their own can also be controlled by the --[no-]allow-shlib-undefined option.
#
#Normally the linker will generate an error message for each reported unresolved symbol but the option --warn-unresolved-symbols can change this to a warning.